include/circt/Support/PrettyPrinterHelpers.h

```diff
@@ -166,6 +166,46 @@ class TokenStringSaver : public PrettyPrinter::Listener {
   void clear() override;
 };
 
+template <typename Callable, typename Data>
+class CallbackSaver {
+
+  /// List of all the unique data associated with each callback token.
+  SmallVector<Data *> list;
+  /// The storage for the callback, as a function object.
+  std::function<void(unsigned)> callbackObj;
+
+  /// Note: Callable class must implement a callable with signature::
+  /// void (Data*)
+public:
+  CallbackSaver() = default;
+  CallbackSaver(Callable &c) { init(c); }
+
+  /// Create and initialize a function object, capturing the callable object and
+  /// a poiner to the data vector. The callable object contains all the values
+  /// shared across the callback tokens. The Data list indexed by the parameter
+  /// provides the unique data for each clalback.
```

Comment:
```suggestion
  /// Create and initialize a function object, capturing the callable object and
  /// a pointer to the data vector. The callable object contains all the values
  /// shared across the callback tokens. The Data list indexed by the parameter
  /// provides the unique data for each callback.
```

FWIW this captures a pointer to the callable (mentioned below), which must live longer than this class and the tokens generated by it.  Hmm.